edi_oca/models/edi_configuration.py

```diff
@@ -0,0 +1,189 @@
+    def _date_to_string(dt, utc=True):
+        if not dt:
+            return ""
+        if utc:
+            dt = to_utc(dt)
+        return fields.Date.to_string(dt)
+
+    @staticmethod
+    def _datetime_to_string(dt, utc=True):
+        if not dt:
+            return ""
+        if utc:
+            dt = to_utc(dt)
+        return fields.Datetime.to_string(dt)
+
+    def _time_utils(self):
+        return {
+            "datetime": safe_eval.datetime,
+            "dateutil": safe_eval.dateutil,
+            "time": safe_eval.time,
+            "utc_now": datetime.datetime.utcnow().isoformat(),
+            "date_to_string": self._date_to_string,
+            "datetime_to_string": self._datetime_to_string,
+            "time_to_string": lambda dt: dt.strftime("%H:%M:%S") if dt else "",
+            "first_of": fields.first,
+        }
+
+    def _get_code_snippet_eval_context(self):
+        """Prepare the context used when evaluating python code
+
+        :returns: dict -- evaluation context given to safe_eval
+        """
+        ctx = {
+            "uid": self.env.uid,
+            "user": self.env.user,
+            "DotDict": DotDict,
+            "conf": self,
+        }
+        ctx.update(self._time_utils())
+        return ctx
+
+    def _evaluate_code_snippet(self, snippet, **render_values):
+        if not self._code_snippet_valued(snippet):
+            return {}
+        eval_ctx = dict(render_values, **self._get_code_snippet_eval_context())
+        safe_eval.safe_eval(snippet, eval_ctx, mode="exec", nocopy=True)
+        result = eval_ctx.get("result", {})
+        if not isinstance(result, dict):
+            return {}
+        return result
+
+    def edi_exec_snippet_before_do(self, record, **kwargs):
+        self.ensure_one()
+        todo = kwargs.get("todo", False)
+        vals = {
+            "todo": todo,
+        }
+        # Execute snippet before do
+        vals_before_do = self._evaluate_code_snippet(
+            self.snippet_before_do, record=record
+        )
+
+        # Prepare data
+        vals.update(
+            {
+                "todo": vals_before_do.get("todo", False),
+                "snippet_do_vars": vals_before_do.get("snippet_do_vars", False),
+                "event_only": vals_before_do.get("event_only", False),
+                "tracked_fields": vals_before_do.get("tracked_fields", False),
+                # Etc ...
+            }
+        )
+        # Update EDIAutoInfo if "snippet_do_vars" in vals
+        # TODO: Add check value in snippet_do_vars
+        if vals.get("snippet_do_vars", False) and todo:
+            for rec in todo:
+                rec.reinitialize(vals["snippet_do_vars"])
+            vals.update({"todo": todo})
+        return vals
+
+    def edi_exec_snippet_do(self, record, **kwargs):
+        self.ensure_one()
+        old_value = kwargs.get("old_vals", {}).get(record.id, False)
+        new_value = kwargs.get("vals", {}).get(record.id, False)
```

Comment:
```suggestion
        old_value = kwargs.get("old_vals", {}).get(record.id, {})
        new_value = kwargs.get("vals", {}).get(record.id, {})
```
In some cases, we want to check the key in dict, it will support for `in` operator